### PR TITLE
Intrepid2 - CMake test generation for different computing platforms

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
## Motivation

This PR generates tests with device and value type based on the active device type in Kokkos. This is for convenience in enabling AMD later. For now, the main objective is to replace existing test suite with the auto-generated tests. For the AMD related issues, I will submit a separate PR.

## Testing

Tested on weaver.

### CODE REVIEW

File: packages/intrepid2/unit-test/Shared/ArrayTools/CMakeLists.txt [hunk 1]
```diff
@@ -1,15 +1,44 @@
...
 IF(Kokkos_ENABLE_CUDA)
-  ADD_SUBDIRECTORY(Cuda)
+  LIST(APPEND Intrepid2_TEST_ETI_DEVICE_NAME "CUDA")
 ENDIF()
+IF(Kokkos_ENABLE_HIP)
+  LIST(APPEND Intrepid2_TEST_ETI_DEVICE_NAME "HIP")
+ENDIF()
+
+FOREACH(ETI_DEVICE_NAME IN LISTS Intrepid2_TEST_ETI_DEVICE_NAME)
+  FOREACH(ETI_FILE IN LISTS Intrepid2_TEST_ETI_FILE)
```
Comment (CamelliaDPG) [inline packages/intrepid2/unit-test/Shared/ArrayTools/CMakeLists.txt:20]: This ETI approach for testing looks like a major improvement for the test build system.  I have not yet understood details of how it works, but I love the fact that we can replace the per-architecture CMakeLists.txt.

**Comment by CamelliaDPG** [comment]
@kyungjoo-kim, could you be more specific about the work that still needs to be done converting execution space to device type parameters?

Also, I don't understand what you're suggesting about DeviceAssert.hpp in #1; I think you're missing a verb in what you typed ("We may want to DeviceAssert.hpp").

File: packages/intrepid2/src/Projection/Intrepid2_ProjectionStruct.hpp [hunk 2]
```diff
@@ -90,16 +90,19 @@ class ProjectionStruct {
   enum EvalPointsType {BASIS, TARGET};
 
   typedef Kokkos::pair<ordinal_type,ordinal_type> range_type;
-  typedef typename Kokkos::Impl::is_space<DeviceType>::host_mirror_space::execution_space HostSpaceType;
-  typedef Kokkos::DynRankView<ValueType,HostSpaceType> view_type;
-  typedef Kokkos::View<range_type**,HostSpaceType> range_tag;
+  /// KK : do we really need this complication instead of using default host exec space ????
```
Comment (mperego) [inline packages/intrepid2/src/Projection/Intrepid2_ProjectionStruct.hpp:93]: I don't think we do. I think we can use the default memory space.

**Comment by kyungjoo-kim** [comment]
@CamelliaDPG I mean to move device assert into the util header file. 

I am not sure if this code belong to @CamelliaDPG or @mperego but I found that some of projection tools uses execution space template parameter.
```
      using ExecutionSpace = typename DeviceType::execution_space;                                                                                                                       
      using ProjectionTools  = Experimental::ProjectionTools<ExecutionSpace>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType                       
      using ProjectionStruct = Experimental::ProjectionStruct<ExecutionSpace,PointScalar>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType    
```

File: packages/intrepid2/src/Projection/Intrepid2_ProjectionStruct.hpp (hunk 2 above)
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Projection/Intrepid2_ProjectionStruct.hpp:93]: @CamelliaDPG Is it fine that I replace  is_space<>::host_execution_space with default host exec/memory space ?

Comment (CamelliaDPG) [reply packages/intrepid2/src/Projection/Intrepid2_ProjectionStruct.hpp:93]: @kyungjoo-kim, yes, I agree.

**Comment by mperego** [comment]
> @CamelliaDPG I mean to move device assert into the util header file. If you agree, then I will merge the two files.
> 
> I am not sure if this code belong to @CamelliaDPG or @mperego but I found that some of projection tools uses execution space template parameter.
> 
> ```
>       using ExecutionSpace = typename DeviceType::execution_space;                                                                                                                       
>       using ProjectionTools  = Experimental::ProjectionTools<ExecutionSpace>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType                       
>       using ProjectionStruct = Experimental::ProjectionStruct<ExecutionSpace,PointScalar>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType    
> ```

I think that projection tools and struct now support the DeviceType as template. So you can pass the DeviceType as a template in the code you pasted above

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG I mean to move device assert into the util header file. If you agree, then I will merge the two files.
> 

Got it; yes, that's fine with me.

> I am not sure if this code belong to @CamelliaDPG or @mperego but I found that some of projection tools uses execution space template parameter.
> 
> ```
>       using ExecutionSpace = typename DeviceType::execution_space;                                                                                                                       
>       using ProjectionTools  = Experimental::ProjectionTools<ExecutionSpace>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType                       
>       using ProjectionStruct = Experimental::ProjectionStruct<ExecutionSpace,PointScalar>; // TODO: when ProjectionTools supports it, replace template argument with DeviceType    
> ```

Got it; thanks for clarifying.  Sounds like probably a search and replace would take care of it?

**Comment by mperego** [comment]
@kyungjoo-kim Thanks for working on this! Are you satified by mine and @CamelliaDPG comments, or do you need more input from us?

**Comment by kyungjoo-kim** [comment]
@CamelliaDPG I already tried and it was not a simple search and replace so I leave it as it is for now.

**Comment by kyungjoo-kim** [comment]
@mperego That's good so far. I just wonder if there is another reason that I do not know when you use a certain approach e.g., is_space and separate device assert. I will add more AMD tests.

**Comment by mperego** [comment]
> @mperego That's good so far. I just wonder if there is another reason that I do not know when you use a certain approach e.g., is_space and separate device assert. I will add more AMD tests.

To be honest, I don't recall why I did that. I might have copied that piece of code from somewhere else and thought it was the right thing to do.

**Comment by mperego** [comment]
@kyungjoo-kim how's this going? Do you have any rough estimate of when it will be completed? thanks.

**Comment by kyungjoo-kim** [comment]
@mperego Within the next 3 months ?

**Comment by mperego** [comment]
> @mperego Within the next 3 months ?

earlier is better. EMPIRE needs to be working with AMD by Q2.

**Comment by kyungjoo-kim** [comment]
@mperego Testing ETI is done. This still needs your fix on Projection tools. You can directly add your fix in this PR branch. Or if you prefer to fix it later, you can just approve this PR so that it can be merged. Let me know which way you prefer. Probably the code will pass the test anyway if kokkoskernels is enabled by default.

**Comment by mperego** [comment]
@kyungjoo-kim thanks! This PR is very useful.
Why don't you enable Cuda tests for projections only when Kokkos Kernel is enabled?

**Comment by kyungjoo-kim** [comment]
@mperego The serial code path has a uvm access problem (real problem). this needs to be fixed. we do not need temporal work around for the problem.

**Comment by kyungjoo-kim** [comment]
@e10harvey Would you help me reproduce the configure error ? I followed the wiki page and I could not load modules on weaver. 
```
[kyukim @weaver11] test > export THOME=/ascldap/users/kyukim/Work/lib/trilinos/kyukim
[kyukim @weaver11] test > source $THOME/cmake/std/sems/PullRequestIntel17.0.1TestingEnv.sh
-bash: /projects/sems/modulefiles/utils/sems-archive-modules-init.sh: No such file or directory
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-gcc/5.3.0'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-intel/17.0.1'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-mpich/3.2'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-boost/1.63.0/base'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-zlib/1.2.8/base'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-hdf5/1.10.6/parallel'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-netcdf/4.7.3/parallel'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-parmetis/4.0.3/parallel'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-scotch/6.0.3/nopthread_64bit_parallel'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-superlu_dist/5.2.2/32bit_parallel'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-cmake/3.17.1'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-ninja_fortran/1.8.2'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-git/2.10.1'
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-archive-python/3.5.2'
```

**Comment by mperego** [comment]
@kyungjoo-kim I think all Trilinos Cuda builds enable Kokkos-Kernels. That's why I didn't find out the issue before. Thanks for identifying the issue and I agree we need to fix the serial path. I'll take care of it in a subsequent PR. However, the serial path is not essential at the moment because we have an alternative which is what is used now by applications like Empire.
To leave things unchanged in terms of test coverage, we need to enable Cuda tests for projections when KokkosKernel is enabled.  Projections are used in Empire and it's not wise to turn off the Cuda tests.

**Comment by kyungjoo-kim** [comment]
@mperego kokkoskernels is an optional package. assuming kokkoskernels is always enabled when cuda is enabled is not a good approach. For the moment, I will do the temporal work around but it needs to be fixed.

**Comment by kyungjoo-kim** [comment]
@e10harvey 

I was able to follow the wiki page on aspic machine. Is the error message okay that I can just overlook or is it something I need to take care ? 
```
[kyukim@ascicgpu047 prtest]$ source $THOME/cmake/std/sems/PullRequestGCC7.2.0TestingEnv.sh
[kyukim@ascicgpu047 prtest]$ cmake -C $THOME/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake -DTrilinos_ENABLE_Intrepid2=ON $THOME
loading initial cache file /ascldap/users/kyukim/Work/lib/trilinos/kyukim/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake
CMake Error: Error processing file: /ascldap/users/kyukim/Work/lib/trilinos/kyukim/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake
```

**Comment by kyungjoo-kim** [comment]
@mperego I applied the derivative dimension commonly for all tests. I see that some PR tests already failed and compare the generated test with the tests already established in the master branch. For HCURL_TET_In SLFadDFad test, it uses dummy things. Is this something you intended to test dummy things ? 
```
constexpr int num_deriv = 0;
constexpr int max_deriv = 1;
#define ConstructWithLabelOutView(obj, ...) obj(#obj, __VA_ARGS__, max_deriv+1)
#define ConstructWithLabelPointView(obj, ...) obj(#obj, __VA_ARGS__, num_deriv+1)
```

**Comment by mperego** [comment]
> @mperego I applied the derivative dimension commonly for all tests. I see that some PR tests already failed and compare the generated test with the tests already established in the master branch. For HCURL_TET_In SLFadDFad test, it uses dummy things. Is this something you intended to test dummy things ?
> 
> ```
> constexpr int num_deriv = 0;
> constexpr int max_deriv = 1;
> #define ConstructWithLabelOutView(obj, ...) obj(#obj, __VA_ARGS__, max_deriv+1)
> #define ConstructWithLabelPointView(obj, ...) obj(#obj, __VA_ARGS__, num_deriv+1)
> ```

We do not have good tests for Fad types. For the most part we check that the code compiles and run but we do not check that the derivatives are correct. I don't recall why I decided to use DFad with 0 derivatives for the point type in this case, but it's a legit value and it should work.
What errors do you see?

**Comment by kyungjoo-kim** [comment]
Simply it got exception when nonzero derivative dimension.

**Comment by mperego** [comment]
> Simply it got exception when nonzero derivative dimension.

OK, I'll look into that as well.

**Comment by kyungjoo-kim** [comment]
@trilinos/framework the pr tester reports that cuda 10 test is failed but i could not check it from the dashboard. How can i check it ?

**Review by mperego — APPROVED** [review]
This is a huge work! Thans for taking care of that Kyungjoo.

**Comment by e10harvey** [comment]
> I was able to follow the wiki page on aspic machine. Is the error message okay that I can just overlook or is it something I need to take care ?
> 
> ```
> [kyukim@ascicgpu047 prtest]$ source $THOME/cmake/std/sems/PullRequestGCC7.2.0TestingEnv.sh
> [kyukim@ascicgpu047 prtest]$ cmake -C $THOME/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake -DTrilinos_ENABLE_Intrepid2=ON $THOME
> loading initial cache file /ascldap/users/kyukim/Work/lib/trilinos/kyukim/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake
> CMake Error: Error processing file: /ascldap/users/kyukim/Work/lib/trilinos/kyukim/cmake/std/PullRequestLinuxGCCX.YTestingSettings.cmake
> ```
> 
> Actually the configure failed after all. It looks like that either the wiki page is outdated or the system has been changed.

@kyungjoo-kim: I think you need to fill in "X.Y" here. For example, for the gcc 8.3.0 build, use "cmake -C [PullRequestLinuxGCC8.3.0TestingSettings.cmake](https://github.com/trilinos/Trilinos/blob/develop/cmake/std/PullRequestLinuxGCC8.3.0TestingSettings.cmake) ...".

**Comment by e10harvey** [comment]
These temporary cuda builds (used until https://github.com/trilinos/Trilinos/issues/10147 is resolved) post to trilinos-cdash.sandia.gov.

**Comment by kyungjoo-kim** [comment]
@e10harvey Ah... that is my mistake. I can try it again. I looked up the trilinos-cdash.sandia.gov. I cannot find this PR test in the dash board but I only see the master branch merge PR test. Could you help me figure out what is going on in the CUDA test ?

**Comment by e10harvey** [comment]
> Could you help me figure out what is going on in the CUDA test ?

Yea, sorry about that. Config & build pass. There are 4 test errors:
Screen Shot 2022-02-28 at 10 12 03 AM

The first test fails with:
```
export TPETRA_ASSUME_CUDA_AWARE_MPI=0; jsrun  '-M -disable_gpu_hooks' '-p' '1' '--rs_per_socket' '4' '/path/to@3/pull_request_test/packages/intrepid2/unit-test/Projection/Intrepid2_unit-test_Projection_test_DeRham_commutativity_HEX_CUDA_DOUBLE.exe' 'PrintItAll'
===============================================================================
|                                                                             |
|              Test 1 (DeRham Commutativity - HGRAD-HCURL)                    |
|                                                                             |
===============================================================================
Considering Hex 0: [ 0 1 2 3 4 5 6 7 ] and Hex 1: [ 4 5 6 7 8 9 10 11 ]
Kokkos::View ERROR: attempt to access inaccessible memory spaceERROR:  One or more process (first noticed rank 0) terminated with signal 6
jsrun return value: 134
```

**Comment by kyungjoo-kim** [comment]
@e10harvey Thanks.

**Comment by mperego** [comment]
@kyungjoo-kim I think projection tools are uve free, however, the DeRahm tests are not (in fact we never run them with cuda). I remember discussing with you and you said that we didn't need all the test to run on Cuda. Can you please enable cuda for the other 8 tests in the projections?

**Comment by kyungjoo-kim** [comment]
@mperego the projection tool unit test is not uvm free. actually the test has never been tested with CUDA space. let's address this later after this PR.

**Comment by mperego** [comment]
These are the projection tests that run with cuda
https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/unit-test/Projection/Cuda/CMakeLists.txt

**Comment by kyungjoo-kim** [comment]
@mperego For now I will surgically instantiation for the unit tests. However, we would like to cover all unit tests without uvm.

**Comment by ZUUL42** [comment]
Here's a link to the Cuda builds for this PR: https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=63&value1=vortex&field2=buildname&compare2=63&value2=PR-9919&field3=buildstamp&compare3=63&value3=Pull%20Request
You can find the URL in the Cuda output, but it's a bit buried. There are plans to change that.

**Comment by kyungjoo-kim** [comment]
@mperego Can you re-approve this to merge ?

**Review by mperego — APPROVED** [review]

**Pull request merged**